src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs

```diff
@@ -1,88 +1,86 @@
+        public VotingManager(FederationManager federationManager, ILoggerFactory loggerFactory, SlotsManager slotsManager, IPollResultExecutor pollResultExecutor,
+            INodeStats nodeStats, DataFolder dataFolder, DBreezeSerializer dBreezeSerializer, ISignals signals)
         {
-            this.federationManager = federationManager;
-            this.keyValueRepo = keyValueRepo;
-            this.slotsManager = slotsManager;
-            this.pollResultExecutor = pollResultExecutor;
+            this.federationManager = Guard.NotNull(federationManager, nameof(federationManager));
+            this.slotsManager = Guard.NotNull(slotsManager, nameof(slotsManager));
+            this.pollResultExecutor = Guard.NotNull(pollResultExecutor, nameof(pollResultExecutor));
+            this.signals = Guard.NotNull(signals, nameof(signals));
+            this.nodeStats = Guard.NotNull(nodeStats, nameof(nodeStats));
 
             this.locker = new object();
             this.votingDataEncoder = new VotingDataEncoder(loggerFactory);
             this.scheduledVotingData = new List<VotingData>();
+            this.pollsRepository = new PollsRepository(dataFolder, loggerFactory, dBreezeSerializer);
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            this.isInitialized = false;
         }
 
         public void Initialize()
         {
-            this.pendingPolls = this.LoadPolls(PendingPollsDbKey);
+            this.pollsRepository.Initialize();
 
-            if (this.pendingPolls == null)
-            {
-                this.logger.LogDebug("No pending polls found in DB, initializing with empty collection.");
+            this.polls = this.pollsRepository.GetAllPolls();
 
-                this.pendingPolls = new List<Poll>();
-                this.SavePolls(this.pendingPolls, PendingPollsDbKey);
-            }
+            this.signals.OnBlockConnected.Attach(this.OnBlockConnected);
+            this.signals.OnBlockDisconnected.Attach(this.OnBlockDisconnected);
 
-            this.finishedPolls = this.LoadPolls(FinishedPollsDbKey);
+            this.nodeStats.RegisterStats(this.AddComponentStats, StatsType.Component, 1200);
 
-            if (this.finishedPolls == null)
-            {
-                this.logger.LogDebug("No finished polls found in DB, initializing with empty collection.");
-
-                this.finishedPolls = new List<Poll>();
-                this.SavePolls(this.finishedPolls, FinishedPollsDbKey);
-            }
+            this.isInitialized = true;
+            this.logger.LogDebug("VotingManager initialized.");
         }
 
         /// <summary>Schedules a vote for the next time when the block will be mined.</summary>
         /// <exception cref="InvalidOperationException">Thrown in case caller is not a federation member.</exception>
         public void ScheduleVote(VotingData votingData)
         {
+            this.EnsureInitialized();
```

Comment:
Generally I don't like this tests everytime `EnsureInitialized` is called, if a component is intiaizable its the responsibility of the implementer to call it

Reply:
This is to prevent component usage when it's not initialized on purpose